### PR TITLE
[14.0][MIG] product_contract

### PR DESCRIPTION
**Pull request by ilyasProgrammer**
based on https://github.com/OCA/contract/pull/652

### CODE REVIEW

**Comment by francesco-ooops** [comment]
@rousseldenis @mourad-ehm @hparfr this is the finished version of https://github.com/OCA/contract/pull/652 , could you approve?

Thanks!

**Comment by francesco-ooops** [comment]
@OCA/project-service-maintainers could you take a look? thanks!

**Comment by rousseldenis** [comment]
@mourad-ehm What do you think about switching to this ?

File: product_contract/models/sale_order.py
Comment (mourad-ehm) [inline product_contract/models/sale_order.py:70]: I suggest to use sale line r.is_contract field rather then r.product_id.is_contract.
In fact, it is more easy to change the value in sale line (by an other module by make this field as not related to product but entered by on-change ). We have some client that sale the same product as contract or a single time (ex cleaning desktop service).

File: product_contract/models/sale_order.py
Comment (mourad-ehm) [inline product_contract/models/sale_order.py:74]: Same remark

File: product_contract/models/sale_order.py
Comment (mourad-ehm) [inline product_contract/models/sale_order.py:34]: Same remark. May be make a method to get line_to_create_contract. In this manner we can inherit it in other module.

File: product_contract/models/sale_order.py
Comment (mourad-ehm) [inline product_contract/models/sale_order.py:38]: r.is_contract

**Comment by francesco-ooops** [comment]
@OCA/project-service-maintainers could someone take a look? thanks!

**Review by mourad-ehm — COMMENTED** [review]
Hi @ilyasProgrammer thanks for this PR. Could you check my proposition ?

**Comment by francesco-ooops** [comment]
Hi @mourad-ehm , what do you say if we implement these improvements after merge?

**Comment by mourad-ehm** [comment]
@francesco-ooops OK for implement this after merge.

**Review by mourad-ehm — APPROVED** [review]

File: contract/tests/test_contract.py
Comment (hparfr) [inline contract/tests/test_contract.py]: are you sure this change is needed ?

Comment (ilyasProgrammer) [reply contract/tests/test_contract.py]: Yes

Comment (hparfr) [reply contract/tests/test_contract.py]: I know you are not the orignal writer of this test and I just read it quickly. 

But, I don't understand the goal of this test.  I don't see the link between `self.contract.write(vals2)` and the following lines.
And why it doesn't work with 2020-01-01 as a start date.

Comment (ilyasProgrammer) [reply contract/tests/test_contract.py]: Related to _onchange_product_id_recurring_info() in product_contract

Comment (leemannd) [reply contract/tests/test_contract.py]: On the long run isn't it best to use freezegun?

You can add the import at the beginning of the file like `from freezegun import freeze_time`

Then add the decorator on the beginning of your method `@freeze_time("2022-01-01 09:00:00")` and then you can let the tested date to its original value `"2020-01-01"`.

Comment (ilyasProgrammer) [reply contract/tests/test_contract.py]: looks like it worked.

Comment (francesco-ooops) [reply contract/tests/test_contract.py]: @hparfr @leemannd good to approve? :)

**Comment by francesco-ooops** [comment]
@rousseldenis good for you? :)

**Comment by rousseldenis** [comment]
@ilyasProgrammer Could you squash ?

**Comment by ilyasProgrammer** [comment]
@rousseldenis done

File: product_contract/models/contract_line.py
Comment (rousseldenis) [inline product_contract/models/contract_line.py]: This is because display_name is the default rec_name. Please remove it in the definition of the model here.

Then, remove also the display_name field.

You should only implement the _compute_display_name() function. That's all.


IMHO, this compute should go into contract module as its code does not depends on something specific here

File: product_contract/models/product_template.py
Comment (rousseldenis) [inline product_contract/models/product_template.py:19]: FYI, I had done this once upon a time:

https://github.com/OCA/server-ux/pull/510

But no success :sweat:

File: product_contract/models/sale_order.py
Comment (rousseldenis) [inline product_contract/models/sale_order.py]:
```suggestion
            self.update({"need_contract_creation": False})
```
 is better for the call stack before the loop.

But, IMHO, you should build a set of ids for False values and update at the end only those ones.

File: product_contract/models/sale_order.py
Comment (rousseldenis) [inline product_contract/models/sale_order.py]: Better in terms of performances to build a set of ids and then return at the end : contract_model.browse(ids)

File: product_contract/models/sale_order.py
Comment (rousseldenis) [inline product_contract/models/sale_order.py]:
```suggestion
                rec.order_line.mapped("contract_id")
```

active is built-in in all queries by the ORM

File: product_contract/models/sale_order.py
Comment (rousseldenis) [inline product_contract/models/sale_order.py]:
```suggestion
        action = self.env["ir.actions.act_window"]._for_xml_id("contract.action_customer_contract")
```

File: product_contract/models/sale_order.py
Comment (rousseldenis) [inline product_contract/models/sale_order.py]:
```suggestion
        action["domain"] = [("contract_line_ids.sale_order_line_id", "in", self.order_line.ids)]
```

You avoid a search (remove the above one)

File: product_contract/models/sale_order_line.py
Comment (rousseldenis) [inline product_contract/models/sale_order_line.py]: Please change to private method (with '_')

File: product_contract/models/sale_order_line.py
Comment (rousseldenis) [inline product_contract/models/sale_order_line.py:133]: I would have changed those fields (is_auto_renew, auto_renew_interval, auto_renew_rule_type) to computed/stored/readonly=False fields in order to be already compliant with further Odoo onchange deprecation.

File: product_contract/models/sale_order_line.py
Comment (rousseldenis) [inline product_contract/models/sale_order_line.py]:
```suggestion
        res = super()._compute_invoice_status()
```

File: product_contract/models/sale_order_line.py
Comment (rousseldenis) [inline product_contract/models/sale_order_line.py]:
```suggestion
        self.filtered("contract_id").update({"invoice_status": "no"})
```

Better for call stack

File: product_contract/tests/test_product.py
Comment (rousseldenis) [inline product_contract/tests/test_product.py]: Change this to classmethod and TransactionCase to SavepointCase

File: product_contract/tests/test_sale_order.py
Comment (rousseldenis) [inline product_contract/tests/test_sale_order.py]: Same as above.

**Review by rousseldenis — CHANGES_REQUESTED** [review]

**Comment by francesco-ooops** [comment]
Thanks a lot @rousseldenis !

File: contract/tests/test_contract.py
Comment (hparfr) [reply contract/tests/test_contract.py]: Ok for me on this point.

**Comment by ilyasProgrammer** [comment]
@rousseldenis Thanks for the review. Updated according to your suggestions.

**Comment by francesco-ooops** [comment]
@rousseldenis when you can :)

**Comment by marcelsavegnago** [comment]
@ilyasProgrammer take a look at this document please. Your commits need tweaking.
https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-14.0

image

In the migration commit you commit everything you did to allow the module to run in the new version. Anything else implements (new and as long as it is justifiable to do it in a migration PR) then you do it in additional commits

**Comment by ilyasProgrammer** [comment]
@marcelsavegnago this migration is based on existing migration PR, and in this case i cant follow instructions exaclty.
Like MIG commit: 8a3c5a2b8f3c9d025c88eda3b77965f259a35395 is there and done by another developer. I cant inject black, isort, prettier commit in between.
All my commits are fixes, imrpovements, refactoring, which was done after migration.

**Comment by marcelsavegnago** [comment]
@ilyasProgrammer Ok.. anyway, I understand that there are some adjustments to commits... for example remove the [14.0] tag from the commit and I believe that instead of [UPD] it should be [IMP], [RFC] or [FIX]..

Here's an example of a commit that seems to need revision..
[[14.0][UPD] updates after review](https://github.com/OCA/contract/pull/933/commits/447629e6e776af70430bd56af075b58c8e4b056d)

Take a look at the history of old commits

**Comment by marcelsavegnago** [comment]
https://www.odoo.com/documentation/14.0/contributing/development/git_guidelines.html?highlight=imp#tag-and-module-name

image

**Comment by marcelsavegnago** [comment]
@ilyasProgrammer I apologize for the persistence, but I suggest removing the tag with the commit version to make it less cluttered in a possible backport to previous versions. Additionally, I understand the message of the latest commit, but the commit message doesn't adequately represent the changes made. If the changes were made to allow for migration, it would have been better to add them in the migration commit. If they were improvements, I believe it is necessary to change the commit message to something like 'Adding improvements' or something similar would be more appropriate than the current message 'Changes after review'.

**Comment by ilyasProgrammer** [comment]
No worries. Good for me to get more familiar with strict OCA standards ). Thank you.

**Comment by francesco-ooops** [comment]
@marcelsavegnago can this be approved now?

@rousseldenis :)

**Comment by marcelsavegnago** [comment]
> @marcelsavegnago can this be approved now?
> 
> @rousseldenis :)

image

**Comment by marcelsavegnago** [comment]
@ilyasProgrammer @francesco-ooops 

Let's go... to keep it standard

I would move the commit [**14.0][FIX] product_contract: pre-commit format code** before the commit **[MIG] product_contract: Migration to 14.0** and would change the commit message to **[IMP] product_contract: black, isort, prettier** 

And I would squash these commits below with the **migration commit**
image

**Comment by ilyasProgrammer** [comment]
updated

File: product_contract/models/contract_line.py
Comment (rousseldenis) [inline product_contract/models/contract_line.py]: To avoid too much sets, update once:

```suggestion
                rec.update({"recurring_rule_type": rec.product_id.recurring_rule_type,
                ...
                })
```

**Review by marcelsavegnago — APPROVED** [review]

**Comment by francesco-ooops** [comment]
@OCA/project-service-maintainers can we merge now?

**Comment by francesco-ooops** [comment]
@rousseldenis :)

File: product_contract/tests/test_sale_order.py
Comment (rousseldenis) [inline product_contract/tests/test_sale_order.py:153]: Is this call necessary as it should be automatic?

Comment (ilyasProgrammer) [reply product_contract/tests/test_sale_order.py:153]: Yes it is necessary

**Comment by ilyasProgrammer** [comment]
updated

**Review by rousseldenis — APPROVED** [review]

**Comment by rousseldenis** [comment]
@ilyasProgrammer Can be merged as is ?

**Comment by ilyasProgrammer** [comment]
yes please

**Comment by francesco-ooops** [comment]
@OCA/project-service-maintainers go? :)

**Comment by francesco-ooops** [comment]
> Hi @ilyasProgrammer thanks for this PR. Could you check my proposition ?

@mourad-ehm feel free to propose an IMP PR to this module :)

**Pull request merged**
